File: 백준/Silver/16173. 점프왕 쩰리 （Small）/점프왕 쩰리 （Small）.js
const input = require("fs")
  .readFileSync("/dev/stdin", "utf8")
  .trim()
  .split("\n");

// 첫번째줄에 N이 주어지고
// N xN 그리드가 주어진다
// 0,0 에서 시작해서 N-1,N-1이면 된다
// 해당 바닥의 숫자만큼 오른쪽 or 아래로만 갈 수 있다
// 마지막칸에만 -1이 나머지는 0-100 정수
// 도달하면 HaruHaru, 안되면 Hing
// bfs로 visited 찍으면서 가면된다.

const solution = (input) => {
  const n = Number(input[0]);
  const grid = [];

  for (let i = 1; i < input.length; i++) {
    grid.push(input[i].split(" ").map((v) => Number(v)));
  }
  const visited = Array(n)
    .fill(0)
    .map((v) => Array(n).fill(false));

  const queue = [[0, 0]];
  visited[0][0] = true;

  while (queue.length) {
    const [x, y] = queue.shift();
    if (grid[x][y] === -1) {
      console.log("HaruHaru");
      return;
    }
    const step = grid[x][y];

    if (x + step < n && !visited[x + step][y]) {
      queue.push([x + step, y]);
      visited[x + step][y] = true;
    }

    if (y + step < n && !visited[x][y + step]) {
      queue.push([x, y + step]);
      visited[x][y + step] = true;
    }
  }

  console.log("Hing");
};

solution(input);

File: 백준/Silver/16173. 점프왕 쩰리 （Small）/점프왕 쩰리 （Small）.py
n = int(input())
grid = []

for i in range(n):
  grid.append(list(map(int, input().split())))

visited = [[False]*n for _ in range(n)]

queue = [[0,0]]
visited[0][0] = True

isArrive = False

while len(queue):
  x,y = queue.pop(0)

  if grid[x][y] == -1:
    isArrive = True
    print('HaruHaru')
    break
  
  step = grid[x][y]

  if y + step < n and not visited[x][y + step]:
    queue.append([x, y + step])
    visited[x][y + step] = True

  if x + step < n and not visited[x+step][y]:
    queue.append([x + step, y])
    visited[x+step][y] = True

if not isArrive:
  print('Hing')